acra-censor/common/logging_logic.go

```diff
@@ -0,0 +1,268 @@
+	return queryWriter.serializationTimeout
+}
+
+// DumpQueries writes all queries into file.
+func (queryWriter *QueryWriter) DumpQueries() error {
+	rawData := queryWriter.serializeQueries(queryWriter.Queries)
+	if err := queryWriter.logStorage.WriteAll(rawData); err != nil {
+		queryWriter.logger.WithError(ErrCantOpenFileError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorIOError).Errorln("Can't dump queries to storage")
+		return err
+	}
+	return nil
+}
+
+// Release dumps all Captured queries to file and resets Captured queries list.
+func (queryWriter *QueryWriter) Release() {
+	queryWriter.DumpQueries()
+	queryWriter.signalBackgroundExit <- true
+	queryWriter.reset()
+}
+
+func (queryWriter *QueryWriter) reset() {
+	queryWriter.Queries = nil
+}
+
+func (queryWriter *QueryWriter) readStoredQueries() error {
+	queryWriter.mutex.Lock()
+	defer queryWriter.mutex.Unlock()
+
+	q, err := queryWriter.deserializeQueries()
+	if err != nil {
+		queryWriter.logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorIOError).Errorln("Can't read stored queries")
+		return err
+	}
+	queryWriter.Queries = q
+	return nil
+}
+
+func (queryWriter *QueryWriter) dumpBufferedQueries() error {
+	queryWriter.mutex.Lock()
+	defer queryWriter.mutex.Unlock()
+	partialRawData := queryWriter.serializeQueries(queryWriter.Queries[queryWriter.queryIndex:])
+	if err := queryWriter.logStorage.Append(partialRawData); err != nil {
+		return err
+	}
+	queryWriter.queryIndex = len(queryWriter.Queries)
+	return nil
+}
+
+func (queryWriter *QueryWriter) deserializeQueries() ([]*QueryInfo, error) {
+	bufferBytes, err := queryWriter.logStorage.ReadAll()
+	if err != nil {
+		return nil, err
+	}
+	var queries []*QueryInfo
+	if len(bufferBytes) != 0 {
+		for _, line := range bytes.Split(bufferBytes, []byte{'\n'}) {
+			if len(line) == 0 {
+				continue
+			}
+			var oneQuery QueryInfo
+			if err = json.Unmarshal(line, &oneQuery); err != nil {
+				return nil, err
+			}
+			queries = append(queries, &oneQuery)
+		}
+	}
+	return queries, nil
+}
+
+func (queryWriter *QueryWriter) serializeQueries(queries []*QueryInfo) []byte {
+	var linesToAppend []byte
+	var tempQueryInfo = &QueryInfo{}
+	for _, queryInfo := range queries {
+		tempQueryInfo.RawQuery = queryInfo.RawQuery
+		tempQueryInfo.IsForbidden = queryInfo.IsForbidden
+		jsonQueryInfo, err := json.Marshal(tempQueryInfo)
+		if err != nil {
+			queryWriter.logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorQuerySerializeError).Errorln("Can't serialize stored queries")
+		}
+		if len(jsonQueryInfo) > 0 {
+			jsonQueryInfo = append(jsonQueryInfo, '\n')
+			linesToAppend = append(linesToAppend, jsonQueryInfo...)
+		}
+	}
+	return linesToAppend
+}
+
+// GetAllInputQueries returns a list of non-masked RawQueries.
+func (queryWriter *QueryWriter) GetAllInputQueries() []string {
+	var queries []string
+	for _, queryInfo := range queryWriter.Queries {
```

Comment:
here we need Rlock. we read shareable resource that may be changed during read. and go runtime will panic on loop through map that was changed at this time.

Reply:
Can you please explain what is Rlock? I googled but only usual Mutex.lock found